Extending `hs.plot.plot_spectra` normalisation functionality

aidanc151 opened the pull request:
### Description of the change
- Extending `hs.plot.plot_spectra` normalise parameter functionality implemented in #3419
- Add ability to normalise by ranges/positions in signal axis
- Add ability to define such ranges/positions by either index position or calibrated units using same syntax as `.isig`
- Add new normalise_type parameter to `hs.plot.plot_spectra` so that in the case when a range is defined, the signals can be normalised to either the `np.mean` or `np.max` of the data in this range. This allows for normalisation to 1) peaks and 2) slopes/non-zero plateaus in the signal, e.g. the Bremsstrahlung background in EDX spectra.

This PR is currently just a docstring, I previously had an old fork with this working for `style = 'overlap'` but will need to make updates to accommodate for the changes in `utils.py`.  I'm not totally convinced the addition of a further parameter is the best way to achieve all the functionality described above. I'm open to suggestions if anyone has ideas for a less clunky/more user friendly way on doing this.

### Progress of the PR
- [x] Update docstring
- [ ] Change implemented for all `style` options
- [ ] Update user guide
- [ ] Add an changelog entry in the `upcoming_changes` folder 
- [ ] Add tests,
- [ ] Ready for review.

aidanc151 committed:
Extended_normalisation_functionality.py

Adding functionality to normalisation parameter in hs.plot.plot_spectra, where signals can be normalised using same syntax as .isig[ ]

ericpre commented:
I agree that this is getting complicated and difficult to explain/understand what it does. What's about supporting passing a function that does the normalisation (range, type of normalisation). The function will take a signal and return an numpy array?

`normalise_type` could be merged with `normalise` to cover cases where the range doesn't need to be specify. Other cases would use custom function passed to `normalise`.

aidanc151 commented:
I'm not sure if I'm understanding correctly, would the normalise function be like an additional wrapper to the signal? Like so: 

`hs.plot.plot_spectra(signal,  normalise = signal.normalise(range = (300, 400), type = 'max')`)

and then you would also have the option to just pass a range or even True value to the normalise parameter without the wrapper: 

`hs.plot.plot_spectra(signal,  normalise = (300, 400))` ?

ericpre commented:
Apologies for coming back on this only now - there were more urgent stuff to sort out and it fell through the cracks, since.
Here is more explanation on what I meant: enable the `normalise` argument to take a user-defined function. For example:

```python
def my_normalise_function(signal):
   min_ =  signal.data.min()
   max_ = signal.data.max()
   return (signal.data - min_) / (max_ - min_)

hs.plot.plot_spectra(signal, normalise=my_normalise_function)
```

The most common normalisation function could be pre-defined and some examples added to the gallery to explain/illustrate what it does.

aidanc151 committed:
Merge branch 'hyperspy:RELEASE_next_minor' into RELEASE_next_minor

aidanc151 committed:
Signal normalisation example 

Demonstrates different custom normalisation functions passed into `hs.plot.plot_spectra`s `normalise` parameter

aidanc151 committed:
Update utils.py

Updated `hs.plot.plot_spectra` docstring
Added condition `to _parse_array` so that if a parameter is a function, then it applies the function to `data` variable. Otherwise, for `normalise = True` default normalisation is applied

aidanc151 commented:
I've started an example under the Data visualization section that shows the application of some custom normalisation functions. I created two unique spectra from `hs.data.luminescence_signal()` for the sake of visualising the different normalisation functions, though my method feels a little messy.

I've implemented a condition in [utils.py](https://github.com/hyperspy/hyperspy/pull/3455/commits/b6a5b388c7e95a59d22287be7528061e7c22a289) so that if a custom function is provided to `hs.plot.plot_spectra`'s  `normalise` parameter then it should be applied to the data.  Now I've played with it a bit, I agree its a more elegant solution without over complicating `hs.plot.plot_spectra`'s parameters

ericpre reviewed (commented):
@aidanc151, this looks very good. It needs a test to cover the new lines. There are some [guidance ](https://hyperspy.org/hyperspy-doc/current/dev_guide/testing.html#running-and-writing-tests) on writing tests in the contributor guide.

The other failures (documentation and pre-commit) in the checks are due to formatting and these are trivial to fix - I am happy to do them later.

Is there a "common" normalisation function that you think would be worth adding? The current `normalise=True` could be changed to `normalised="min_max"` and other could be added if there are straightforward.

aidanc151 committed:
Merge branch 'hyperspy:RELEASE_next_minor' into RELEASE_next_minor

aidanc151 committed:
Add tests for extending spectra normalisation functionality

aidanc151 commented:
I've adapted the existing normalisation test to take True, False, and a callable function, I created an arbitrary normalisation function in the test file for this callable parameter.

When I run pytest locally on specifically the test file, it seems to pass everything. Although I've noticed noticed that the hyperspy/tests/drawing/plot_signal1D folder doesn't export .png's for the True, False, and a callable function parameters I set, so I expect something isn't correct. I will have another go in the next days

aidanc151 committed:
Responses to testing errors

Response to formatting error and to resolve .png's not being outputted

aidanc151 committed:
Refactor section headers in signal_normalisation.py

Updated section headers for better clarity and consistency in the signal normalisation script.

aidanc151 commented:
> Is there a "common" normalisation function that you think would be worth adding? The current `normalise=True` could be changed to `normalised="min_max"` and other could be added if there are straightforward.

@ericpre I would've suggested adding a normalise by a single position in the signal axis as an option (e.g. a peak), but this would require to somehow pass a value in the signal axis into `hs.plot.plot_spectra` as well - which we have avoided with this whole approach of adding a callable `normalise` parameter. 

For this criteria of a normalisation function that doesn't require any additional signal axis value(s), maybe a function that normalises by the next most intensity peak, though I'm unsure how complicated this may become.

ericpre committed:
Fix formatting, docstring and tests

ericpre commented:
Thank you @aidanc151 for finishing this PR. I tweak the documentation and the test. The test failure has to do plot testing where images of the plot are compared with baseline images and here some were missing of the update to the test. This is what I have fixed to get the test suite to pass on Azure Pipeline.

Regarding the "common" or standard normalisation, I was asking just in case, that there is a simple or established function that you had in mind, but it doesn't need to be one, so this is why I tidied up and merged!
